Normalize the Navigation block appender behavior between canvas and list view contexts

getdave opened the pull request:
## What

This PR aligns the behavior of Navigation block appenders between canvas and list view contexts. Currently, the list view appender uses a different insertion flow than the canvas appender, leading to inconsistent user experience.

## Why

The Navigation block's list view and canvas contexts should provide consistent appender behavior. Users expect the same insertion experience regardless of whether they're editing in the canvas or using the list view sidebar. This alignment improves the overall user experience and reduces confusion.

This should open the door for improved UX in https://github.com/WordPress/gutenberg/pull/71192.

## How

- **Enhanced List View Appender**: Modified the list view appender to use the same block insertion flow as the canvas appender
- **Unified Block List Settings**: Both contexts now respect the same `directInsert` and `defaultBlock` settings configured in the Navigation block
- **Consistent LinkUI Integration**: List view appender now properly integrates with the Navigation block's LinkUI component for link editing
- **Aligned Cleanup Behavior**: Both contexts now handle block cleanup consistently when editing is cancelled or completed

## Known follow ups

- We should change the `+` inserter tooltip to be `Add page` in _both_ List View and canvas.
- We need to fix/update the e2e tests.


## Testing

1. **Canvas Context**: Verify that Navigation block appender still works as expected with direct insertion
2. **List View Context**: Verify that list view appender provides the same insertion experience
3. **LinkUI Integration**: Test that both contexts properly open LinkUI for newly inserted blocks
4. **Block Cleanup**: Ensure both contexts handle empty/incomplete blocks consistently

## Screenshots




https://github.com/user-attachments/assets/f3c7cfe1-3369-4215-ae9d-671ddb48909f



## Key Changes from Trunk

- List view appender now respects Navigation block's `directInsert` and `defaultBlock` settings
- List view appender properly integrates with Navigation block's LinkUI component
- Consistent block cleanup behavior between canvas and list view contexts
- Unified user experience for Navigation block editing across both contexts

getdave committed:
refactor: eliminate prop drilling for directInsert in PrivateListView

- Add directInsert to ListViewContext to eliminate prop drilling
- Update ListViewBranch to use context instead of props
- Update Appender to use context instead of props
- Maintain directInsert prop at PrivateListView level to set context value
- Cleaner component interfaces and centralized configuration

getdave committed:
Use block list settings api

getdave committed:
Use original const def

getdave committed:
Access setting directly in appender

getdave committed:
Revert branch component to match trunk

getdave committed:
Revert another file to trunk

getdave committed:
Pass entire block to callback

getdave committed:
Revert changes to menu controls

getdave commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -386,9 +386,7 @@ export default compose( [
-					onSelectOrClose( {
-						clientId: blockToInsert?.clientId,
-					} );
+					onSelectOrClose( blockToInsert );
```
This will be backwards compat as the new object will still contain `clientId`.

getdave commented on packages/block-library/src/navigation/constants.js:
```diff
@@ -1,5 +1,6 @@
 export const DEFAULT_BLOCK = {
 	name: 'core/navigation-link',
+	attributes: { type: 'page', kind: 'post-type' },
```
This is needed to ensure the directly inserted block is the correct _variation_ (i.e. `Page`) for Nav blocks.

getdave committed:
Properly handle cleanup if no link selected

getdave committed:
Allow for setting blocks inserterd by Add block Quick Inserter

getdave commented:
@scruffian I'm currently stuck on this problem:

The issue is that because we `directInsert` the `core/navigation/page` block variation it remains in the canvas when we select a custom block using the `Add block` (via the `QuickInserter`). We need to clean up like we do in canvas.

scruffian commented:
> The issue is that because we directInsert the core/navigation/page block variation it remains in the canvas when we select a custom block using the Add block (via the QuickInserter). We need to clean up like we do in canvas.

Is it possible to show a list of pages, without inserting the block first? That seems cleaner than having to do the cleanup...

getdave committed:
fix: clean up original nav link block when selecting different block in list view

When using the 'Add block' feature in the Navigation block's list view context,
the original auto-inserted navigation link block was not being cleaned up when
a different block was selected via the QuickInserter. This caused duplicate
blocks to remain in the list view.

The fix adds a wrapper function handleSetInsertedBlock that removes the
original inserted block before setting the new selected block, ensuring
only the newly selected block remains in the list view.

This aligns the list view behavior with the canvas context behavior.

getdave committed:
fix: prevent focus stealing when selecting blocks in Navigation list view

When selecting a different block from the QuickInserter in the Navigation block's
list view context, the original auto-inserted block was being removed with
automatic block selection enabled. This caused focus stealing that closed the
list view and switched to the canvas.

The fix adds preventBlockSelection = false to removeBlock() calls to prevent
automatic block selection, keeping the list view open and maintaining proper
focus behavior.

getdave commented on packages/block-editor/src/components/list-view/appender.js:
```diff
@@ -21,6 +21,16 @@ export const Appender = forwardRef(
+				return settings?.directInsert || false;
+			},
+			[ clientId ]
+		);
```
This will cause ALL list views that have an inserter to mirror the behaviour set in the block list settings. This makes sense to me and I believe is limited to the Nav block as it's the only one able to consume the PrivateListView.

Confidence check appreciated.

talldan replied:
Aligning how it works makes sense to me. Not sure if we could move towards them using the same component/hook or if they're too fundamentally different.

Only code quality / performance feedback is that this `useSelect` could be merged with the one a few lines below to reduce the number of store subscriptions.

getdave committed:
refactor: rename setInsertedBlock prop to onBlockInsert for better clarity

Rename the setInsertedBlock prop to onBlockInsert across the Navigation block
components to make the prop name more generic and descriptive of its purpose
as a callback function that's called when a block is inserted.

This improves code readability and makes the prop name more reusable across
different contexts while maintaining all existing functionality.

scruffian commented:
> Block Cleanup: Ensure both contexts handle empty/incomplete blocks consistently

This works when I click in the sidebar to cancel link creation, but not when I click in the canvas:

https://github.com/user-attachments/assets/4048adaa-373f-4025-a5a4-82dfad817ce8

getdave commented:
This appears to be an issue relating to the canvas iframe. The Link UI is outside the iframe and thus the focusoutside logic doesn't seem to detect clicks within the editor canvas. 

If you click anywhere else in the editor UI _other than_ the editor canvas the cleanup works fine. It's just selecting stuff inside the canvas.

At this point I don't have a solution but essentially the `onClose` of `Popover` isn't firing due to internals of `Popover`.

getdave committed:
fix: add iframe-aware cleanup for LinkUI in Navigation block List View

- Add useEditorCanvasCleanup hook to handle iframe interaction cleanup
- Fixes issue where auto-inserted Navigation Link blocks weren't cleaned up when clicking on canvas blocks in List View context
- Uses standardized iframe selector approach consistent with rest of codebase
- Hook automatically manages event listener lifecycle based on LinkUI visibility

getdave committed:
fix: add iframe-aware cleanup for LinkUI in Navigation block List View

- Add useEditorCanvasCleanup hook to handle iframe interaction cleanup
- Fixes issue where auto-inserted Navigation Link blocks weren't cleaned up when clicking on canvas blocks in List View context
- Uses standardized iframe selector approach consistent with rest of codebase
- Hook automatically manages event listener lifecycle based on LinkUI visibility

getdave commented on packages/block-library/src/navigation/edit/menu-inspector-controls.js:
```diff
@@ -34,15 +35,97 @@ const BLOCKS_WITH_LINK_UI_SUPPORT = [
+ * @param {Object}   insertedBlock  - The currently inserted block data
+ * @param {Function} cleanupFn      - Function to call for cleanup when iframe is clicked
+ */
+function useEditorCanvasCleanup( isLinkUIOpen, insertedBlock, cleanupFn ) {
```
@ciampo I had to add this manual detect for clicks within the editor canvas because useDialog's focus detection wasn't firing `onClose` on `Popover` for clicks within the `iframe`. I wonder if you've seen a problem like this before. If not should we be looking to fix in `useDialog`?

Please see https://github.com/WordPress/gutenberg/pull/71163#issuecomment-3188321391

getdave replied:
@scruffian This is the fix for [the issue you noticed](https://github.com/WordPress/gutenberg/pull/71163#issuecomment-3188321391) where clicking inside the canvas didn't trigger the close of the Link UI in the Navigation block's "List View" sidebar.

scruffian replied:
It works!

scruffian replied:
Looks like a similar problem to https://github.com/WordPress/gutenberg/issues/58825

getdave replied:
It is the same issue. You can replicate it by:

- select a random block in the editor canvas (not the Nav block)
- `wp.data.select('core/block-editor').getSelectedBlockClientId()` - make a note of the ID
- manually select the Nav block and open List View and cliock the appender - see the Nav Link "Page" variation block inserted and the Link UI appears
- back in Dev tools console issue the following to _programmatically_ select a different block - `wp.data.dispatch('core/block-editor').selectBlock( CLIENT_ID_FROM_ABOVE )`
- see that the onClose handler doesn't run because the Nav Item "Page" that was directly inserted is not "cleaned up"

This proves that it is the unmount behaviour causing the issue. It's exactly the same as described in https://github.com/WordPress/gutenberg/issues/58825#issuecomment-2111103844. Basically the timeout which would trigger the `onClose` handler is being cancelled when the `useFocusOutside` hook unmounts as a result of its consuming component unmounting. The unmount happens because a different block gets selected and this removes the Nav block's List View and thus unmounts everything.

This is essentially a bug with `Popover` so we have to work around it.

jeryj replied:
I've been debugging this. I'm not finished yet, but it seems related to the iframe swallowing up focus events. I think if you programmatically move focus to an element inside the iframe, then I think the issue is events not properly bubbling up past the iframe boundary.

Also, a reason I think it's not the unmount behavior, is that you can follow your same example from a popover in the style guide sidebar for a color and it also doesn't close the popover even though component doesn't unmount.

ciampo replied:
Trying to catch up after being AFK — out of curiosity, to test @jeryj 's theory, if you render a "scrim" element behind the popover / on top of the iframe (so that the iframe doesn't swallow the focus events), would that fix it?

I remember dealing with a related issue (ie. `iframe` swallowing pointer events) in [this PR](https://github.com/WordPress/gutenberg/pull/55149) — the description and the screencasts should explain the issue that I had encountered.

jeryj replied:
When I was testing, I was trying to send focus to elements via the console when focus was inside the popover. Turns out, when you do `el.focus()` from the console, no blur events or focus events are fired :/ They do get fired when you programmatically do it inside the code on the page, but not via the console 😞  So, my method of testing wasn't helpful.

> if you render a "scrim" element behind the popover / on top of the iframe (so that the iframe doesn't swallow the focus events), would that fix it?

I think if this worked, it would only solve this for click events. That might be enough to fix their bug though.

scruffian commented on packages/block-editor/src/components/list-view/appender.js:
```diff
@@ -79,7 +89,7 @@ export const Appender = forwardRef(
 					isAppender
 					selectBlockOnInsert={ false }
-					shouldDirectInsert={ false }
+					shouldDirectInsert={ directInsert }
```
I guess we don't have any instances where this is false anymore.

getdave replied:
It should mirror the block list settings.

getdave committed:
Fix test to expect directly inserted page link

getdave committed:
Fix another test to expect direct inserted Page block

getdave committed:
tweak comments

getdave committed:
Remove hacky fix to onClose failure

getdave committed:
Revert alteration to hook

getdave commented on packages/compose/src/hooks/use-focus-outside/index.ts:
```diff
@@ -87,10 +89,13 @@ export default function useFocusOutside(
 	useEffect( () => {
+		if ( ! cancelOnUnmount ) {
+			return;
+		}
```
This line causes problems with the `onClose` callback of `<Popover>` not being called if it's parent component unmounts. The issue is that the timeout is cancelled by this `useEffect` on unmount which results in the `onClose` not being called.

In this PR the result of this is the bug which @scruffian identified and which I have documented in https://github.com/WordPress/gutenberg/pull/71163#discussion_r2278741629.

talldan replied:
Is the issue is that you always want to trigger the callback whenever the popover closes or unmounts, so that `cleanupInsertedBlock` is run?

I think `cancelOnUnmount` might be quite a difficult API to understand when and when not to use.

I personally wonder if it's a bug that `onClose` doesn't get called when the popover/dialog unmounts. Changing it now might cause back compat issues, especially if an `event` can't be provided, so that's probably not an option.

Alternative might be to have an `onUnmount` callback that always gets called, or maybe if the component that unmounts is one you can modify (`AdditionalBlockContent`?) you could add an unmount effect there that calls `cleanupInsertedBlock`.

It'd be good to get wider feedback (from component contributors) on the change if it's needed.

edit: oh wait, I see there's discussion below, not sure how I missed it.

getdave commented:
This PR is BLOCKED by a bug in `Popover` (its sub hooks) which means the `onClose` callback which removes directly inserted Nav Link blocks is not fired if the Nav block is _unmounted_ whilst Link UI is open.

I've proposed a potential route in https://github.com/WordPress/gutenberg/pull/71163#discussion_r2278741629 but I think we need input from @WordPress/gutenberg-components team on the route forward.

aduth commented:
> I've proposed a potential route in [#71163 (comment)](https://github.com/WordPress/gutenberg/pull/71163#discussion_r2278741629) but I think we need input from @WordPress/gutenberg-components team on the route forward.

A few initial impressions:

- Should this be default behavior of `Popover` rather than a prop?
- The timeout behavior was originally added in #3235 in-part to normalize inconsistencies with `FocusEvent#relatedTarget` specifically related to IE11, so I'd wonder if we could get rid of the timeout altogether
- If that seems like a worthwhile route to consider, I wouldn't necessarily expect it to have to happen here, but maybe we could at least hide the `cancelOnUnmount` behavior as a [private API](https://developer.wordpress.org/block-editor/contributors/code/coding-guidelines/#private-function-arguments) to avoid it being a publicly-available option in the meantime.

talldan commented on packages/block-library/src/navigation-link/link-ui.js:
```diff
@@ -206,9 +207,12 @@ function UnforwardedLinkUI( props, ref ) {
 		<Popover
 			ref={ ref }
 			placement="bottom"
```
I noticed in the video that the positioning of the popover is a bit awkward. Not sure if there's a way to improve it. 🤔

getdave replied:
Yes agreed. We'll need to tweak that.

getdave commented:
I'd love to unblock this one. Thanks for all the work @jeryj and @aduth. 

I wonder how far off we are from resolving https://github.com/WordPress/gutenberg/pull/71252?

getdave committed:
Optimise useSelect calls

getdave commented on packages/block-library/src/navigation/edit/leaf-more-menu.js:
```diff
@@ -146,7 +146,6 @@ export default function LeafMoreMenu( props ) {
 						<AddSubmenuItem
 							block={ block }
 							onClose={ onClose }
-							expanded
```
The expanded prop was being passed to AddSubmenuItem but never actually used in the component. The function only uses expandedState and expand props for its logic. This was a cleanup to remove dead code that was likely leftover from an earlier implementation.

getdave committed:
Simplify cleanup function to avoid passing functions around

getdave commented:
@jeryj I've approved https://github.com/WordPress/gutenberg/pull/71252. Hopefully we can bring it in and then rebase this one.

getdave committed:
Fix mess up from rebase

scruffian reviewed (approved):
This is working well now. A much better UX too, let's bring it in...
